SpatialGDK/Source/SpatialGDK/Private/Interop/ClientNetLoadActorHelper.cpp

```diff
@@ -0,0 +1,142 @@
+	{
+		if (UObject* DynamicSubObject = NetDriver->PackageMap->GetObjectFromNetGUID(*SubObjectNetGUID, false))
+		{
+			NetDriver->PackageMap->ResolveSubobject(DynamicSubObject, ObjectRef);
+			UE_LOG(LogClientNetLoadActorHelper, Verbose,
+				   TEXT("Found reusable dynamic SubObject (ObjectRef offset: %u) for ClientNetLoad actor with entityId %d"),
+				   ObjectRef.Offset, ObjectRef.Entity);
+			return DynamicSubObject;
+		}
+	}
+	return nullptr;
+}
+
+void FClientNetLoadActorHelper::EntityRemoved(const Worker_EntityId EntityId, const AActor* Actor)
+{
+	ClearDynamicSubObjectRefs(EntityId);
+	if (USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(EntityId))
+	{
+		for (UObject* DynamicSubObject : Channel->CreateSubObjects)
+		{
+			FNetworkGUID SubObjectNetGUID = NetDriver->PackageMap->GetNetGUIDFromObject(DynamicSubObject);
+			if (SubObjectNetGUID.IsValid())
+			{
+				FUnrealObjectRef SubObjectRef = NetDriver->PackageMap->GetUnrealObjectRefFromNetGUID(SubObjectNetGUID);
+				if (SubObjectRef.IsValid() && IsDynamicSubObject(*NetDriver, Actor, SubObjectRef.Offset))
+				{
+					SaveDynamicSubObjectRef(SubObjectRef, SubObjectNetGUID);
+					UE_LOG(
+						LogClientNetLoadActorHelper, Verbose,
+						TEXT("Saved reusable dynamic SubObject ObjectRef (ObjectRef offset: %u) for ClientNetLoad actor with entityId %d"),
+						SubObjectRef.Offset, SubObjectRef.Entity);
+				}
+			}
+		}
+	}
+}
+
+FNetworkGUID* FClientNetLoadActorHelper::GetSavedDynamicSubObjectNetGUID(const FUnrealObjectRef& ObjectRef)
+{
+	if (TMap<FUnrealObjectRef, FNetworkGUID>* EntityMappings = DynamicSubObjectRefToGuid.Find(ObjectRef.Entity))
+	{
+		if (FNetworkGUID* NetGUID = EntityMappings->Find(ObjectRef))
+		{
+			return NetGUID;
+		}
+	}
+	return nullptr;
+}
+
+void FClientNetLoadActorHelper::SaveDynamicSubObjectRef(const FUnrealObjectRef& ObjectRef, const FNetworkGUID& NetGUID)
+{
+	TMap<FUnrealObjectRef, FNetworkGUID>& EntityMappings = DynamicSubObjectRefToGuid.FindOrAdd(ObjectRef.Entity);
+	EntityMappings.Emplace(ObjectRef, NetGUID);
+}
+
+void FClientNetLoadActorHelper::ClearDynamicSubObjectRefs(const Worker_EntityId InEntityId)
+{
+	if (TMap<FUnrealObjectRef, FNetworkGUID>* EntityMappings = DynamicSubObjectRefToGuid.Find(InEntityId))
+	{
+		for (auto DynamicSubObjectIterator = EntityMappings->CreateIterator(); DynamicSubObjectIterator; ++DynamicSubObjectIterator)
+		{
+			DynamicSubObjectIterator.RemoveCurrent();
+		}
+	}
+}
+
+void FClientNetLoadActorHelper::RemoveRuntimeRemovedComponents(const Worker_EntityId EntityId, const TArray<ComponentData>& NewComponents)
+{
+	auto ContainedInComponentsArr = [this, EntityId, &NewComponents](const FUnrealObjectRef& CheckComponentObjRef) {
+		for (const ComponentData& Component : NewComponents)
+		{
+			// Skip if this isn't a generated component
+			if (Component.GetComponentId() < SpatialConstants::STARTING_GENERATED_COMPONENT_ID
+				&& Component.GetComponentId() != SpatialConstants::DORMANT_COMPONENT_ID)
```

Comment:
hmmm? why this dormant component?

Reply:
can't remember, was some optimisation i copied in(along with the starting generated comp check I think), will remove it